packages/contracts/test/upgrade/dao.ts

```diff
@@ -0,0 +1,400 @@
+let daoV120Proxy: DAO_V1_2_0;
+
+let daoV100Implementation: string;
+let daoV120Implementation: string;
+let daoCurrentImplementaion: DAO;
+
+const EMPTY_DATA = '0x';
+
+const DUMMY_METADATA = ethers.utils.hexlify(
+  ethers.utils.toUtf8Bytes('0x123456789')
+);
+
+const FORWARDER_1 = `0x${'1'.repeat(40)}`;
+const FORWARDER_2 = `0x${'2'.repeat(40)}`;
+
+describe('DAO Upgrade', function () {
+  before(async function () {
+    signers = await ethers.getSigners();
+
+    // We don't use the typchain here but directly grab the artifacts. This will be changed in an upcoming PR again.
+    DAO_V1_0_0 = new DAO_V1_0_0__factory(signers[0]);
+    DAO_V1_2_0 = new DAO_V1_2_0__factory(signers[0]);
+    DAO_Current = new DAO__factory(signers[0]);
+
+    // Deploy the v1.3.0 implementation
+    daoCurrentImplementaion = await DAO_Current.deploy();
+  });
+
+  context(`v1.0.0 to v1.3.0`, function () {
+    beforeEach(async function () {
+      daoV100Proxy = await deployWithProxy<DAO_V1_0_0>(DAO_V1_0_0);
+      await daoV100Proxy.initialize(
+        DUMMY_METADATA,
+        signers[0].address,
+        ethers.constants.AddressZero,
+        daoExampleURI
+      );
+
+      // Store the v1.0.0 implementation
+      daoV100Implementation = await readImplementationValueFromSlot(
+        daoV100Proxy.address
+      );
+
+      // Grant the upgrade permission
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        UPGRADE_PERMISSIONS.UPGRADE_DAO_PERMISSION_ID
+      );
+    });
+
+    it('does not corrupt the DAO storage', async () => {
+      // Upgrade and call `initializeFrom`.
+      const upgradeTx = await daoV100Proxy.upgradeToAndCall(
+        daoCurrentImplementaion.address,
+        DAO_Current.interface.encodeFunctionData('initializeFrom', [
+          [1, 0, 0],
+          EMPTY_DATA,
+        ])
+      );
+
+      // Check the stored implementation.
+      const implementationAfterUpgrade = await readImplementationValueFromSlot(
+        daoV100Proxy.address
+      );
+      expect(implementationAfterUpgrade).to.equal(
+        daoCurrentImplementaion.address
+      );
+      expect(implementationAfterUpgrade).to.not.equal(daoV100Implementation);
+
+      // Check the emitted implementation.
+      const emittedImplementation = (
+        await findEventTopicLog(upgradeTx, DAO_V1_0_0.interface, 'Upgraded')
+      ).args.implementation;
+      expect(emittedImplementation).to.equal(daoCurrentImplementaion.address);
+
+      // Check that storage is not corrupted.
+      expect(await daoV100Proxy.callStatic.daoURI()).to.equal(daoExampleURI);
+    });
+
+    it('does not corrupt permissions', async () => {
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        ethers.utils.id('EXECUTE_PERMISSION')
+      );
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        ethers.utils.id('ROOT_PERMISSION')
```

Comment:
Isn't root already granted to this testing wallet?

Reply:
Correct, fixed in 649732bf29643735916cd16e3ecf397ca6992afd